cpp/osh_bool_stat.h

```diff
@@ -18,26 +19,101 @@ bool isatty(Str* fd_str, word_t* blame_word);
 inline bool DoUnaryOp(Id_t op_id, Str* s) {
   mylib::Str0 path(s);
 
-  // TODO: also call lstat(), etc.
-  struct stat st;
-  if (stat(path.Get(), &st) < 0) {
-    return false;
-  }
-  auto mode = st.st_mode;
-
-  switch (op_id) {
-  // synonyms for existence
-  case Id::BoolUnary_a:
-  case Id::BoolUnary_e:
-    return true;
+  const char *zPath = path.Get();
 
-  case Id::BoolUnary_f:
-    return S_ISREG(mode);
+  if (op_id == Id::BoolUnary_h ||
+      op_id == Id::BoolUnary_L )
+  {
+    struct stat st;
+    if (lstat(zPath, &st) < 0) {
+      return false;
+    }
 
-  case Id::BoolUnary_k:
-    return (mode & S_ISVTX) != 0;
+    return S_ISLNK(st.st_mode);
+  }
+  else
+  {
+    struct stat st;
+    if (stat(zPath, &st) < 0) {
+      return false;
+    }
+
+    auto mode = st.st_mode;
+
+    switch (op_id) {
+    // synonyms for existence
+    case Id::BoolUnary_a:
+    case Id::BoolUnary_e:
+      return true;
+
+    case Id::BoolUnary_s:
+      return st.st_size != 0;
+
+    case Id::BoolUnary_d:
+      return S_ISDIR(mode);
+
+    case Id::BoolUnary_f:
+      return S_ISREG(mode);
+
+    case Id::BoolUnary_k:
+      return (mode & S_ISVTX) != 0;
+
+    case Id::BoolUnary_p:
+      return S_ISFIFO(mode);
+
+    case Id::BoolUnary_O:
+      return st.st_uid == geteuid();
+
+    case Id::BoolUnary_G:
+      return st.st_gid == getegid();
+
+    case Id::BoolUnary_u:
+      return mode & S_ISUID;
+
+    case Id::BoolUnary_g:
+      return mode & S_ISGID;
+
+
+    // NOTE(Jesse): This implementation MAY have a bug.  On my system (Ubuntu
+    // 20.04) it returns a correct result if the user is root (elevated with
+    // sudo) and no execute bits are set for a file.
+    //
+    // A bug worked around in the python `posix` module here is that the above
+    // (working) scenario is not always the case.
+    //
+    // https://github.com/python/cpython/blob/8d999cbf4adea053be6dbb612b9844635c4dfb8e/Modules/posixmodule.c#L2547
+    //
+    // As well as the dash source code found here (relative to this repo root):
+    //
+    // _cache/spec-bin/dash-0.5.10.2/src/bltin/test.c
+    // See `test_file_access()`
+    //
+    // We could also use the `stat` struct to manually compute the permissions,
+    // as shown in the above `test.c`, though the code is somewhat obtuse.
+    //
+    // There is further discussion of this issue in:
+    // https://github.com/oilshell/oil/pull/1168
+    //
+    // And a bug filed for it at:
+    //
+    // https://github.com/oilshell/oil/issues/1170
+    //
+      case Id::BoolUnary_x:
+        return faccessat(AT_FDCWD, zPath, X_OK, AT_EACCESS) == 0;
```

Comment:
Hm the indentation here looks a little funny, it can be fixed with

`test/lint.sh format-oil` 


which runs `clang-format` on all the C++ code.  (I will put this in the cheat sheet)

But I'll just merge this now since we can do it at any time.